### PR TITLE
[VideoPlayer] WinRenderer: re-work video rendering

### PR DESCRIPTION
**Pull request by afedchin**
* This adds LUT and dithering support for windows,
* Re-factor render buffers
* Re-factor rendering methods.

@FernetMenta please review in part of work with buffers

### CODE REVIEW

File: xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.h [hunk 1]
```diff
@@ -24,7 +24,12 @@
 #include "lcms2.h"
 #endif
 
-#include <string>
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.h]: causes build errors

Comment (afedchin) [inline xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.h]: oops. thanks

File: xbmc/video/dialogs/GUIDialogCMSSettings.cpp [hunk 2]
```diff
@@ -20,7 +20,7 @@
 
 #include "system.h"
 
-#ifdef HAS_GL
+#if defined(HAS_GL) || defined(HAS_DX)
```
Comment (FernetMenta) [inline xbmc/video/dialogs/GUIDialogCMSSettings.cpp]: we should do this with cmake, only include where needed.

File: xbmc/cores/VideoPlayer/VideoRenderers/WinRenderBuffer.cpp [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+
+#define PLANE_Y 0
+#define PLANE_U 1
+#define PLANE_V 2
+#define PLANE_UV 1
+#define PLANE_D3D11 0
+
+CRenderBuffer::CRenderBuffer()
+  : loaded(false)
+  , frameIdx(0)
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/WinRenderBuffer.cpp]: ctor initialization is hard to maintain. much better is c++11 syle by assignment in header
your code, your decision

File: xbmc/cores/VideoPlayer/VideoRenderers/WinRenderer.cpp [hunk 4]
```diff
@@ -911,38 +979,28 @@ bool CWinRenderer::RenderCapture(CRenderCapture* capture)
 //********************************************************************************************************
 // YV12 Texture creation, deletion, copying + clearing
 //********************************************************************************************************
-void CWinRenderer::DeleteYV12Texture(int index)
+void CWinRenderer::DeleteRenderBuffer(int index)
 {
   CSingleLock lock(g_graphicsContext);
   ReleaseBuffer(index);
-  SAFE_DELETE(m_VideoBuffers[index]);
+  m_renderBuffers[index].Release();
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/WinRenderer.cpp]: that is exaclty what line 985 does, isn't it?

**Review by FernetMenta — APPROVED** [review]
this is a rather big change and I don't claim to understand the details. I looked at the buffer handling as requested. looks good

File: xbmc/cores/VideoPlayer/VideoRenderers/WinRenderer.cpp (hunk 4 above)
Comment (afedchin) [inline xbmc/cores/VideoPlayer/VideoRenderers/WinRenderer.cpp]: it's not the same. `ReleaseBuffer` releases video buffer only. `Release` method releases render buffer resources. perhaps need to change name for better understanding.

File: xbmc/video/dialogs/GUIDialogCMSSettings.cpp (hunk 2 above)
Comment (afedchin) [inline xbmc/video/dialogs/GUIDialogCMSSettings.cpp]: yes, but I'm not original author so I left this as is

Comment (FernetMenta) [inline xbmc/video/dialogs/GUIDialogCMSSettings.cpp]: the define was here before cmake. now things can be improved when they are touched anyway.

**Comment by afedchin** [comment]
@FernetMenta please review https://github.com/xbmc/xbmc/pull/12475/commits/f98e9ef4f5223482af71e35798e236bf4053b4a9 because you are the author of an idea of surface context

**Comment by FernetMenta** [comment]
@afedchin yes, that makes sense. An alternative for MarkRender/MarkDecode is to grab a reference on the ffmpeg buffer in CDecoder::Decode(). Then ffmpeg won't return the view. See VTB for an example:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Video/VTB.cpp#L236

**Comment by afedchin** [comment]
@FernetMenta thanks. with this approach dxva pool will look much simpler

**Comment by afedchin** [comment]
build error is unrelated.

**Comment by MartijnKaijser** [comment]
All green :)

**Comment by jchapple** [comment]
When using this with an Intel NUC7i7BN on a non-HDR TV to decode 2160 24hz HDR content, using a generic `bt2020to709-4000to300-0.15blacklevel-0offset` 3DLUT file, the resulting video image is completely distorted, green and noisy.

The same content with the same 3DLUT file works well for HDR->SDR on a different rig with a GTX 970.

**Pull request merged**